waku/v2/peermanager/peer_manager.go

```diff
@@ -120,3 +109,119 @@ func (pm *PeerManager) pruneInRelayConns() {
+	//Assuming all addresses have peerId
+	info, err := peer.AddrInfoFromP2pAddr(address)
+	if err != nil {
+		return "", err
+	}
+
+	//Add Service peers to serviceSlots.
+	for _, proto := range protocols {
+		pm.AddPeerToServiceSlot(proto, info.ID, origin)
+	}
+
+	//Add to the peer-store
+	err = pm.addPeer(info.ID, info.Addrs, origin)
+	if err != nil {
+		return "", err
+	}
+
+	return info.ID, nil
+}
+
+// RemovePeer deletes peer from the peerStore after disconnecting it.
+// It also removes the peer from serviceSlot.
+func (pm *PeerManager) RemovePeer(peerID peer.ID) {
+	pm.host.Peerstore().RemovePeer(peerID)
+	//Search if this peer is in serviceSlot and if so, remove it from there
+	// TODO:Add another peer which is statically configured to the serviceSlot.
+	for proto, peers := range pm.serviceSlots {
+		for i, peer := range peers {
+			if peer == peerID {
+				pm.serviceSlots[proto][i] = ""
+			}
+		}
+	}
+}
+
+// AddPeerToServiceSlot adds a peerID to serviceSlot.
+// Adding to peerStore is expected to be already done by caller.
+// If relay proto is passed, it is not added to serviceSlot.
+func (pm *PeerManager) AddPeerToServiceSlot(proto protocol.ID, peerID peer.ID, origin wps.Origin) {
+	if proto == WakuRelayIDv200 {
+		pm.logger.Warn("Cannot add Relay peer to service peer slots")
+		return
+	}
+
+	//For now adding the peer to serviceSlot which means the latest added peer would be given priority.
+	//TODO: Ideally we should sort the peers per service and return best peer based on peer score or RTT etc.
+	pm.logger.Info("Adding peer to service slots", logging.HostID("peer", peerID), zap.String("service", string(proto)))
+	pm.serviceSlots[proto] = append(pm.serviceSlots[proto], peerID)
+}
+
+// SelectPeer is used to return a random peer that supports a given protocol.
+// If a list of specific peers is passed, the peer will be chosen from that list assuming
+// it supports the chosen protocol, otherwise it will chose a peer from the service slot.
+// If a peer cannot be found in the service slot, a peer will be selected from node peerstore
+func (pm *PeerManager) SelectPeer(proto protocol.ID, specificPeers []peer.ID, logger *zap.Logger) (peer.ID, error) {
+	// @TODO We need to be more strategic about which peers we dial. Right now we just set one on the service.
+	// Ideally depending on the query and our set  of peers we take a subset of ideal peers.
+	// This will require us to check for various factors such as:
+	//  - which topics they track
+	//  - latency?
+
+	filteredPeers, err := utils.FilterPeersByProto(pm.host, specificPeers, proto)
+	if err != nil {
+		return "", err
+	}
+	if proto == WakuRelayIDv200 {
+		return utils.SelectRandomPeer(filteredPeers, pm.logger)
+	}
+
+	//Try to fetch from serviceSlot
+	peerIDs, ok := pm.serviceSlots[proto]
+	if ok || len(peerIDs) > 0 {
+		pm.logger.Info("Got peer from service slots", logging.HostID("peer", peerIDs[0]))
+		return peerIDs[0], nil
```

Comment:
Maybe this should be random here as well? :thinking:

Reply:
I don't see multiple static peers per service being added as of now, hence left it as is. 
Anyways instead of random, at a later point we will have scoring and other params based selection.